File: objc/Chapter1/HitList/ViewController.m
//
//  ViewController.m
//  HitList
//
//  Created by Jinwoo Kim on 6/26/21.
//

#import "ViewController.h"
#import <CoreData/CoreData.h>
#import "AppDelegate.h"

@interface ViewController () <UITableViewDataSource>
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property NSMutableArray *people;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.people = [@[] mutableCopy];
    
    self.title = @"The List";
    [self.tableView registerClass:[UITableViewCell self] forCellReuseIdentifier:@"Cell"];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    // 1
    AppDelegate *appDelegate = (AppDelegate *)UIApplication.sharedApplication.delegate;
    
    if (![appDelegate isKindOfClass:[AppDelegate class]]) {
        return;
    }
    
    NSManagedObjectContext *managedContext = appDelegate.persistentContainer.viewContext;
    
    // 2
    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] initWithEntityName:@"Person"];
    
    // 3
    NSError * _Nullable error = nil;
    
    self.people = [[managedContext executeFetchRequest:fetchRequest error:&error] mutableCopy];
    
    if (error) {
        NSLog(@"Could npt fetch, %@, %@", error, error.userInfo);
    }
}

- (IBAction)addName:(UIBarButtonItem *)sender {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"New Name"
                                                                   message:@"Add a new name"
                                                            preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *saveAction = [UIAlertAction actionWithTitle:@"Save"
                                                         style:UIAlertActionStyleDefault
                                                       handler:^(UIAlertAction * _Nonnull action) {
        
        UITextField * _Nullable textField = alert.textFields.firstObject;
        NSString * _Nullable nameToSave = textField.text;
        
        if (nameToSave == nil) return;
        
        [self saveName:nameToSave];
        [self.tableView reloadData];
    }];
    
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
                                                           style:UIAlertActionStyleCancel
                                                         handler:^(UIAlertAction * _Nonnull action) {}];
    
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {}];
    [alert addAction:saveAction];
    [alert addAction:cancelAction];
    
    [self presentViewController:alert animated:YES completion:^{}];
}

- (void)saveName:(NSString *)name {
    AppDelegate *appDelegate = (AppDelegate *)UIApplication.sharedApplication.delegate;
    
    if (![appDelegate isKindOfClass:[AppDelegate class]]) {
        return;
    }
    
    // 1
    NSManagedObjectContext *managedContext = appDelegate.persistentContainer.viewContext;
    
    // 2
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"Person"
                                              inManagedObjectContext:managedContext];
    
    NSManagedObject *person = [[NSManagedObject alloc] initWithEntity:entity
                                       insertIntoManagedObjectContext:managedContext];
    
    // 3
    [person setValue:name forKey:@"name"];
    
    // 4
    NSError * _Nullable error = nil;
    [managedContext save:&error];
    
    if (error) {
        NSLog(@"Could npt fetch, %@, %@", error, error.userInfo);
    } else {
        [self.people addObject:person];
    }
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.people.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSManagedObject *person = self.people[indexPath.row];
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    cell.textLabel.text = (NSString *)[person valueForKeyPath:@"name"];
    return cell;
}

@end

File: swift/Chapter1/HitList/ViewController.swift
//
//  ViewController.swift
//  HitList
//
//  Created by Jinwoo Kim on 6/21/21.
//

import UIKit
import CoreData

class ViewController: UIViewController {
    @IBOutlet weak var tableView: UITableView!
    
    var people: [NSManagedObject] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        title = "The List"
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        // 1
        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else {
            return
        }
        
        let managedContext = appDelegate.persistentContainer.viewContext
        
        // 2
        let fetchRequest = NSFetchRequest<NSManagedObject>(entityName: "Person")
        
        // 3
        do {
            people = try managedContext.fetch(fetchRequest)
        } catch let error as NSError {
            print("Could not fetch. \(error), \(error.userInfo)")
        }
    }
    
    @IBAction func addName(_ sender: UIBarButtonItem) {
        let alert = UIAlertController(title: "New Name",
                                      message: "Add a new name",
                                      preferredStyle: .alert)
        
        let saveAction = UIAlertAction(title: "Save",
                                       style: .default) { [unowned self] action in
            
            guard let textField = alert.textFields?.first,
                  let nameToSave = textField.text else {
                      return
                  }
            
            self.save(name: nameToSave)
            self.tableView.reloadData()
        }
        
        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel)
        
        alert.addTextField()
        
        alert.addAction(saveAction)
        alert.addAction(cancelAction)
        
        present(alert, animated: true)
    }
    
    func save(name: String) {
        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else {
            return
        }
        
        // 1
        let managedContext: NSManagedObjectContext = appDelegate.persistentContainer.viewContext
        
        // 2
        let entity: NSEntityDescription = NSEntityDescription.entity(forEntityName: "Person", in: managedContext)!
        
        let person: NSManagedObject = NSManagedObject(entity: entity, insertInto: managedContext)
        
        // 3
        person.setValue(name, forKey: "name")
        
        do {
            // 4
            try managedContext.save()
            people.append(person)
        } catch let error as NSError {
            print("Could not save. \(error), \(error.userInfo)")
        }
    }
}

extension ViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return people.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let person = people[indexPath.row]
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        cell.textLabel?.text = person.value(forKeyPath: "name") as? String
        return cell
    }
}
